Add cuda_custom_kernel in DETA

SangbumChoi opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @pacman100

Integrations:

- deepspeed: HF Trainer/Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc and @younesbelkada

Documentation: @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->
@amyeroberts Hi, can you check this PR? This will make training DETA with custom_cuda_kernel same in Deformable DETR!
I also confirmed that training with this kernel working properly in custom dataset

SangbumChoi committed:
enable graident checkpointing in DetaObjectDetection

SangbumChoi committed:
fix missing part in original DETA

SangbumChoi committed:
make style

SangbumChoi committed:
make fix-copies

SangbumChoi committed:
Revert "make fix-copies"

This reverts commit 4041c86c29248f1673e8173b677c20b5a4511358.

SangbumChoi committed:
remove fix-copies of DetaDecoder

SangbumChoi committed:
enable swin gradient checkpointing

SangbumChoi committed:
fix gradient checkpointing in donut_swin

SangbumChoi committed:
add tests for deta/swin/donut

SangbumChoi committed:
Revert "fix gradient checkpointing in donut_swin"

This reverts commit 1cf345e34d3cc0e09eb800d9895805b1dd9b474d.

SangbumChoi committed:
change supports_gradient_checkpointing pipeline to PreTrainedModel

SangbumChoi committed:
Revert "add tests for deta/swin/donut"

This reverts commit 6056ffbb1eddc3cb3a99e4ebb231ae3edf295f5b.

SangbumChoi committed:
Revert "Revert "fix gradient checkpointing in donut_swin""

This reverts commit 24e25d0a14891241de58a0d86f817d0b5d2a341f.

SangbumChoi committed:
Simple revert

SangbumChoi committed:
enable deformable detr gradient checkpointing

SangbumChoi committed:
Merge branch 'main' of https://github.com/SangbumChoi/transformers into fix_deta

SangbumChoi committed:
add gradient in encoder

SangbumChoi committed:
Merge branch 'main' of https://github.com/SangbumChoi/transformers into fix_deta

SangbumChoi committed:
add cuda_custom_kernel function in MSDA

SangbumChoi committed:
make style and fix input of DetaMSDA

SangbumChoi committed:
make fix-copies

SangbumChoi committed:
remove n_levels in input of DetaMSDA

amyeroberts reviewed (commented):
Thanks for adding this - looks great! 

Just some small structural comments

SangbumChoi commented on src/transformers/models/deta/load_custom.py:
@amyeroberts this is copied from deformable detr should I move the modeling_deta.py?

amyeroberts replied:
Ah, I see. So in other models, we keep this in the modeling file e.g. like [here for yoso](https://github.com/huggingface/transformers/blob/0507e69d34f8902422eb4977ec066dd6bef179a0/src/transformers/models/yoso/modeling_yoso.py#L53). I think this is better and more in-line with our one file per-model policy.

SangbumChoi replied:
Understood!

SangbumChoi replied:
@amyeroberts I changed like as yoso format! I can change deformable detr also if it is necessary.

amyeroberts replied:
Yes please - that would be great!

SangbumChoi committed:
minor changes

SangbumChoi committed:
refactor custom_cuda_kernel like yoso format
https://github.com/huggingface/transformers/blob/0507e69d34f8902422eb4977ec066dd6bef179a0/src/transformers/models/yoso/modeling_yoso.py#L53

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_28989). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

amyeroberts reviewed (approved):
Amazing work - thanks for adding!
